kroxylicious/src/test/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandlerTest.java

```diff
@@ -0,0 +1,224 @@
+                assertEquals("embedded", String.valueOf(ctx.srcAddress()));
+                assertEquals("embedded", ctx.clientAddress());
+            }
+            assertEquals(sendApiVersions ? "foo" : null, ctx.clientSoftwareName());
+            assertEquals(sendApiVersions ? "1.0.0" : null, ctx.clientSoftwareVersion());
+            if (saslOffloadConfigured && sendSasl) {
+                assertEquals("alice", ctx.authorizedId());
+            }
+            else {
+                assertNull(ctx.authorizedId());
+            }
+
+            ctx.initiateConnect(CLUSTER_HOST, CLUSTER_PORT, new KrpcFilter[0]);
+            return null;
+        }).when(filter).selectServer(valueCapture.capture());
+
+        var handler = new KafkaProxyFrontendHandler(filter, dp, false, false) {
+            @Override
+            ChannelFuture getConnect(String remoteHost, int remotePort, Bootstrap b) {
+                // This is ugly... basically the EmbeddedChannel doesn't seem to handle the case
+                // of a handler creating an outgoing connection and ends up
+                // trying to re-register the outbound channel => IllegalStateException
+                // So we override this method to short-circuit that
+                outboundChannel = new EmbeddedChannel();
+                return new DefaultChannelPromise(outboundChannel).setSuccess();
+            }
+        };
+        inboundChannel.pipeline().addLast(handler);
+        inboundChannel.pipeline().fireChannelActive();
+
+        assertEquals(State.START, handler.state());
+
+        if (sslConfigured) {
+            // Simulate the SSL handler
+            inboundChannel.pipeline().fireUserEventTriggered(new SniCompletionEvent(SNI_HOSTNAME));
+        }
+
+        if (haProxyConfigured) {
+            // Simulate the HA proxy handler
+            inboundChannel.writeInbound(new HAProxyMessage(HAProxyProtocolVersion.V1,
+                    HAProxyCommand.PROXY, HAProxyProxiedProtocol.TCP4,
+                    "1.2.3.4", "5.6.7.8", 65535, CLUSTER_PORT));
+        }
+
+        if (sendApiVersions) {
+            // Simulate the client doing ApiVersions
+            writeRequest(ApiVersionsRequestData.HIGHEST_SUPPORTED_VERSION, new ApiVersionsRequestData()
+                    .setClientSoftwareName("foo").setClientSoftwareVersion("1.0.0"));
+            assertEquals(State.API_VERSIONS, handler.state());
+            verify(filter, never()).selectServer(handler);
+        }
+
+        if (sendSasl) {
+            if (saslOffloadConfigured) {
+
+                // Simulate the KafkaAuthnHandler having done SASL offload
+                inboundChannel.pipeline().fireUserEventTriggered(new AuthenticationEvent("alice", Map.of()));
+
+                // If the pipeline is configured for SASL offload (KafkaAuthnHandler)
+                // then we assume it DOESN'T propagate the SASL frames down the pipeline
+                // and therefore no backend connection happens
```

Comment:
This comment is valuable but its a bit hidden.

Personally I'd probably have added an explicit test something like `shouldNotPropagateSASLFramesWhenOffloadEnabled()` I tend to think of the tests as an executable specification so I value making the business logic assertions explicit.